### PR TITLE
Streamline object-builder text parsing

### PR DESCRIPTION
**Pull request by plypaul**
### Description

When parsing the Jinja template in where-filters, there is currently a set of factory and factory-specific classes to implement a specific behavior. I've noticed that there are multiple implementations of those classes (if you add the ones in this repo and dependent repos), and it seemed like there was an opportunity to streamline the structure to handle the update for saved-queries and other changes down the road.

This PR adds a `QueryItemDescription` that encapsulates what the user has specified in the text input, and also adds a `QueryItemTextProcessor` that allows you to:

* Find all the `QueryItemDescription`s in a template.
* Render a template for a given `QueryItemDescription` using a configurable format.
* Configure what builder methods (e.g. `.descending()`) are allowed in which cases.

These features seem like it would support existing use cases, but tagging @DevonFulcher to confirm. 

Also, please view by commit. The first commit includes the interface, so if there are issues there, it would be better to get those resolved before updating the later commits.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: dbt_semantic_interfaces/parsing/text_input/rendering_helper.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        args = (time_granularity_name, date_part_name, descending)
+        assert sum(1 for arg in args if arg is not None) == 1, f"Expected exactly 1 argument set, but got {args}"
+
+        builder_method: Optional[ObjectBuilderMethod]
+        if time_granularity_name is not None:
+            builder_method = ObjectBuilderMethod.GRAIN
+        elif date_part_name is not None:
+            builder_method = ObjectBuilderMethod.DATE_PART
+        elif descending is not None:
+            builder_method = ObjectBuilderMethod.DESCENDING
```
Comment (DevonFulcher) [inline dbt_semantic_interfaces/parsing/text_input/rendering_helper.py:193]: Can we somehow remove the necessity for this block? I think we should strive to reduce the amount of boilerplate necessary when adding a builder method.

File: dbt_semantic_interfaces/parsing/text_input/valid_method.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    DEFAULT_MAPPING_FOR_ORDER_BY: ValidMethodMapping = {
+        QueryItemType.METRIC: frozenset({ObjectBuilderMethod.DESCENDING}),
+        QueryItemType.ENTITY: frozenset({ObjectBuilderMethod.DESCENDING}),
+        QueryItemType.DIMENSION: frozenset(
+            {ObjectBuilderMethod.DESCENDING, ObjectBuilderMethod.GRAIN, ObjectBuilderMethod.DATE_PART}
+        ),
+        QueryItemType.TIME_DIMENSION: frozenset({ObjectBuilderMethod.DESCENDING}),
+    }
+
+    DEFAULT_MAPPING: ValidMethodMapping = {
```
Comment (DevonFulcher) [inline dbt_semantic_interfaces/parsing/text_input/valid_method.py:24]: I'm a bit confused by this. Are these the allowed methods for group-by & where-filters?

File: dbt_semantic_interfaces/parsing/text_input/rendering_helper.py (hunk 1 above)
Comment (plypaul) [reply dbt_semantic_interfaces/parsing/text_input/rendering_helper.py:193]: Yeah, this block can be removed. However, I don't think net boiler plate will change - did you have a particular approach in mind?

Comment (plypaul) [reply dbt_semantic_interfaces/parsing/text_input/rendering_helper.py:193]: Yeah, this block can be removed. However, I don't think net boiler plate will change - did you have a particular approach in mind?

File: dbt_semantic_interfaces/parsing/text_input/valid_method.py (hunk 2 above)
Comment (plypaul) [reply dbt_semantic_interfaces/parsing/text_input/valid_method.py:24]: Yeah, that's what it seems like from the implementation in DSI.

File: dbt_semantic_interfaces/parsing/text_input/rendering_helper.py (hunk 1 above)
Comment (plypaul) [reply dbt_semantic_interfaces/parsing/text_input/rendering_helper.py]: Added a commit with that update.

File: dbt_semantic_interfaces/parsing/text_input/valid_method.py (hunk 2 above)
Comment (plypaul) [reply dbt_semantic_interfaces/parsing/text_input/valid_method.py:24]: e.g. https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/dbt_semantic_interfaces/parsing/where_filter/where_filter_time_dimension.py

File: dbt_semantic_interfaces/parsing/text_input/rendering_helper.py (hunk 1 above)
Comment (DevonFulcher) [reply dbt_semantic_interfaces/parsing/text_input/rendering_helper.py]: The change you made is great thank you! Now if we add another builder method, we don't have to add another `elif`.

File: dbt_semantic_interfaces/parsing/text_input/valid_method.py (hunk 2 above)
Comment (DevonFulcher) [reply dbt_semantic_interfaces/parsing/text_input/valid_method.py:24]: Got it thanks. I'm unsure if these mappings are correct then. Does it make sense to have `descending` in group-by & where-filter? Similarly, does it make sense to have `grain` in order-by? That may be worth reconsidering.

**Review by DevonFulcher — APPROVED** [review]
Looks great thank you!

File: dbt_semantic_interfaces/parsing/text_input/valid_method.py (hunk 2 above)
Comment (plypaul) [reply dbt_semantic_interfaces/parsing/text_input/valid_method.py:24]: > Does it make sense to have descending in group-by & where-filter?

To clarify, are you suggesting that we allow descending in group-by & where-filter? Same thing for the grain.

Comment (DevonFulcher) [reply dbt_semantic_interfaces/parsing/text_input/valid_method.py:24]: I'm not sure what is supported right now, but I was just thinking logically, does it make sense for users to define `descending` & `grain` in those use cases?

**Pull request merged**